### PR TITLE
Lambda and Stream APIs

### PR DESCRIPTION
**Pull request by aayushi93**
Ticket URL - https://www.notion.so/Lambda-Stream-APIs-12c724669aea4f48b8528b62c72583ea

### CODE REVIEW

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Override
+    public void process() throws IOException {
+        List<String> file_list = new ArrayList<>();
+        for (File f : listFiles(getRootPath())) {
+            for (String lines : readLines(f)) {
+                if (containsPattern(lines)) {
+                    file_list.add(lines);
+                }
+
```
Comment (zhenzhangca) [inline core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java:34]: remove extra empty lines

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrep.java [hunk 2]
```diff
@@ -1,4 +1,67 @@
-package PACKAGE_NAME;
+package ca.jrvs.apps.grep;
```
Comment (zhenzhangca) [inline core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrep.java:1]: I think this is not "Lambda and Stream APIs", it's Java Grep implementation, right?

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
+    }
+
+
+    @Override
+    public void process() throws IOException {
+        List<String> file_list = new ArrayList<>();
+        for (File f : listFiles(getRootPath())) {
```
Comment (zhenzhangca) [inline core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java]: rootPath is a member field of this class, no need to use getRootPath() here, use rootPath directly.

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            BufferedReader brd = new BufferedReader(new FileReader(inputFile));
+            String l = brd.readLine();
+            while (brd.ready() && l != null) {
+                numLines.add(l);
+                l = brd.readLine();
+            }
+            brd.close();
+        } catch (IOException ioe) {
+            ioe.printStackTrace();
+
```
Comment (zhenzhangca) [inline core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java]: extra empty line

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        }
+        return numLines;
+    }
+
+    @Override
+    public boolean containsPattern(String line) {
+        return Pattern.compile(regex).matcher(line).find();
```
Comment (zhenzhangca) [inline core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java:77]: How about "return line.matches(regex)",  I am not sure, maybe it's Edward's rule.

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrep.java (hunk 2 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrep.java:1]: This particular file is JavaGrep interface which is used for both implementations - JavaGrepImp and JavaGrepLambda

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 5 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java:77]: String.matches() needs entire string to be match with the regex whereas pattern.matcher() allows substring.

Comment (zhenzhangca) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java:77]: Ok, thanks for letting me know that. I don't use regex a lot.

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 1 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java:34]: Done, thanks!

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 3 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java]: Updated!
Thanks.

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 4 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java]: Removed!

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 1 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java:34]: Done, Thanks!

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 3 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java]: Updated!

File: core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java (hunk 4 above)
Comment (aayushi93) [reply core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java]: Removed, thanks!

**Review by zhenzhangca — APPROVED** [review]

**Pull request merged**